src/ray/core_worker/core_worker.cc

```diff
@@ -2759,13 +2759,16 @@ void CoreWorker::HandleExit(const rpc::ExitRequest &request, rpc::ExitReply *rep
                             rpc::SendReplyCallback send_reply_callback) {
   bool own_objects = reference_counter_->OwnObjects();
   // Fail the request if it owns any object.
-  if (own_objects) {
-    reply->set_success(false);
-  } else {
-    reply->set_success(true);
-    Exit(rpc::WorkerExitType::INTENDED_EXIT);
-  }
-  send_reply_callback(Status::OK(), nullptr, nullptr);
+  reply->set_success(!own_objects);
+  send_reply_callback(Status::OK(),
+                      [own_objects, this]() {
+                        // If it doesn't own objects, we'll exit it
+                        if (!own_objects) {
+                          Exit(rpc::WorkerExitType::INTENDED_EXIT);
+                        }
+                      },
+                      // We need to kill it if grpc failed.
+                      [this]() { Exit(rpc::WorkerExitType::INTENDED_EXIT); });
```

Comment:
@stephanie-wang  @rkooo567  let me know whether this makes sense or not.

Reply:
I'd prefer not to include this change; it seems a bit complicated. It would be best if the raylet can handle this type of error.

Reply:
Yeah also @stephanie-wang mentioned we shouldn’t be using this success failure callback, so let’s not do this for now. (It is status quo in terms of network failures with other parts)

Reply:
Raylet has handled this I think. But the error is because of a bug right? What we should do is:

- wait until send_reply succeeded
- exit in the process

There are two places where the error could be:
1. Raylet failed to send: in this case, raylet remove this worker and this worker will be leaked.
2. core worker failed to reply: in this case, 1) raylet remove this worker; 2) this worker will exit on itself.

If we don't ensure the ordering here, there will be a race condition: a) the worker is sending reply; b) the worker is exiting. 

If b) finished first, a) will return an error to raylet. With the current implementation, it'll go through 2) mentioned above.

But this actually is not an error and enforce ordering will make it work, then why don't we just avoid it and the error msg in raylet will be more accurate.

Reply:
I think send_reply is synchronous (please check https://github.com/ray-project/ray/pull/15411#discussion_r616961156), so order is already ensured?

Reply:
@rkooo567 it calls this function which from the comment it got sent asynchronously.
```
  /// Tell gRPC to finish this request and send reply asynchronously.
  void SendReply(const Status &status) {
    state_ = ServerCallState::SENDING_REPLY;
    response_writer_.Finish(reply_, RayStatusToGrpcStatus(status), this);
  }
```

Reply:
I've run it locally for a while and it has iterated for more than 700 iterations. I feel it should be good?

Reply:
I'm not sure if I fully understand all the cases here, but I would still prefer that we did not need to rely on the ordering of the worker's execution here. I think that this is just a more robust solution.

Just to clarify, is the ordering here part of the fix for the original bug about scheduling to dead workers? If not, then I think we should remove this change.

Reply:
Resolved offline with @iycheng. I didn't realize this was also one of the original causes of the bug.